pkg/puller/puller.go

```diff
@@ -104,235 +101,131 @@ func (p *Puller) manage(warmupTime time.Duration) {
-
-			// we defer the actual start of syncing to get out of the iterator first
-			var (
-				peersToSync       []peer
-				peersToRecalc     []peer
-				peersDisconnected = make(map[string]peer)
-			)
 
 			p.syncPeersMtx.Lock()
 
-			// make a map of all peers we're syncing with, then remove from it
-			// the entries we get from kademlia  in the iterator, this way we
-			// know which peers are no longer there anymore (disconnected) thus
-			// should be removed from the syncPeer bin.
-			for po, bin := range p.syncPeers {
-				for peerAddr, v := range bin {
-					peersDisconnected[peerAddr] = peer{addr: v.address, po: uint8(po)}
+			// peersDisconnected is used to mark and prune peers that are no longer connected.
+			peersDisconnected := make(map[string]*syncPeer)
+			for _, bin := range p.syncPeers {
+				for addr, peer := range bin {
+					peersDisconnected[addr] = peer
 				}
 			}
 
-			// EachPeerRev in this case will never return an error, since the content of the callback
-			// never returns an error. In case in the future changes are made to the callback in a
-			// way that it returns an error - the value must be checked.
-			_ = p.topology.EachPeerRev(func(peerAddr swarm.Address, po uint8) (stop, jumpToNext bool, err error) {
+			neighborhoodDepth := p.topology.NeighborhoodDepth()
+			syncRadius := p.reserveState.GetReserveState().StorageRadius
+
+			_ = p.topology.EachPeerRev(func(addr swarm.Address, po uint8) (stop, jumpToNext bool, err error) {
 				if po >= neighborhoodDepth {
-					bp := p.syncPeers[po]
-					// delete from peersDisconnected since we'd like to sync
-					// with this peer
-					delete(peersDisconnected, peerAddr.ByteString())
-
-					// within depth, sync everything
-					if _, ok := bp[peerAddr.ByteString()]; !ok {
-						// we're not syncing with this peer yet, start doing so
-						bp[peerAddr.ByteString()] = newSyncPeer(peerAddr, p.bins)
-						peersToSync = append(peersToSync, peer{addr: peerAddr, po: po})
-					} else {
-						// already syncing, recalc
-						peersToRecalc = append(peersToRecalc, peer{addr: peerAddr, po: po})
+					// add peer to sync
+					if _, ok := p.syncPeers[po][addr.ByteString()]; !ok {
+						p.syncPeers[po][addr.ByteString()] = newSyncPeer(addr, po, p.bins)
 					}
+					// remove from disconnected list as the peer is still connected
+					delete(peersDisconnected, addr.ByteString())
+				} else {
+					// outside of neighborhood, prune peer
+					p.disconnectPeer(addr.ByteString(), po)
 				}
 
-				// if peer is outside of depth, do nothing here, this
-				// will cause the peer to stay in the peersDisconnected
-				// map, leading to cancelling of its running syncing contexts.
-
 				return false, false, nil
 			}, topology.Filter{Reachable: true})
 
-			p.syncPeersMtx.Unlock()
-
-			for _, v := range peersToSync {
-				p.syncPeer(ctx, v.addr, v.po, syncRadius)
+			for addr, peer := range peersDisconnected {
+				p.disconnectPeer(addr, peer.po)
 			}
```

Comment:
Is this going to call `disconnectPeer` for the second time, where first time is on line 133 `p.disconnectPeer(addr.ByteString(), po)`?

Reply:
yes, disconnect for out of depth peers, and disconnect call for peers that were in the syncPeers map but are no longer connected (eg missing in the iterator).